plasmapy/formulary/parameters.py

```diff
@@ -468,7 +474,7 @@ def ion_sound_speed(
 @particles.particle_input
 def thermal_speed(
     T: u.K,
-    particle: particles.Particle = "e-",
+    particle: Particle = "e-",
```

Comment:
```suggestion
    particle: Particle,
```

The test associated with this will need to be updated.

Reply:
I was going to leave this optional, because if the `mass` argument is specified, then the `particle` argument is unused. But of course having the `particle` argument sometimes ignored is very different than giving it a default value.

I'll change this function, too.

Thinking about this more, does it still make since to keep the `mass` argument for this function? I'm probably missing something, but overriding the mass is not something that seems like a common, or expected thing to do. I guess it doesn't hurt to leave it.

Reply:
See also `formulary.collisions._replaceNanVwithThermalV` for a case where the `mass` argument is specified, but now I have to add something like `particle='e-'` for the code to work. But in the context of that function it seems strange to specify a particle.

Maybe an alternative to having the `mass` parameter in the `thermal_speed` function would be to just use a `CustomParticle` object? Something like
```
particle = CustomParticle(mass=m, charge=0)
v = thermal_speed(T, particle)
```

Reply:
I get you.  In that case I think we should drop the `particle` argument altogether and make `mass` a required argument, but reinforce in the examples that you can do `thermal_speed(5*u.ev, mass=Particle("e").mass)`.  A function should always require what it needs to run and having `particle` and `mass` has optional keywords does not satisfy that.

@namurphy What do you think is the best way to proceed?

Reply:
Hm...I prefer the approach of using `CustomParticle`, though this will eventually require finishing #807.  Maybe that's the next issue I should work on...

Reply:
Maybe for #807 you can make `@particle_input` check for `AbstractParticle`? Isn't that the advantage of having an abstract base class... so that any checks for that will still work with concrete subclasses?

Reply:
Yep, that's pretty close to what I was thinking.  We'd also need to make sure that `DimensionlessParticle` does not included, since pretty much everything in `plasmapy.formulary` requires dimensional particles.  We could have a hierarchy like `AbstractParticle` → `DimensionalParticle` → `Particle` & `CustomParticle`.

Reply:
Idk, I'm still not convinced about this.  I feel requiring a user to do `thermal_speed(5 * u.eV, particle=CustomeParticle(mass=5. * u.g))` is just too far away from `thermal_speed(5 * u.eV, mass=5. * u.g)` and will cause user frustration.  It's not as intuitive as `thermal_speed(5 * u.eV, mass=Particle("e").mass)`.  I don't really like either scenario though.

I think it might be cooler to have a decorator do something like...

```python
@from_particle(mass={"grab": mass})
themal_speed(T: u.K, mass: u.g):
    ...
```

Then when doing `thermal_speed(5*u.eV, mass="e")` the decorator knows to grab `Particle("e").mass` and pass that in.  Additionally, `thermal_speed(5 * u.eV, mass=5. * u.g)` would still work as expected.

Reply:
In my mind, there will be more times that I want the thermal speed of a particle than I will want the thermal speed for a given mass. I think that the point of having a package like this is to make it easy to do the common things.
```python
thermal_speed(T, "e-")
```
is easier than
```python
thermal_speed(T, Particle("e-").mass)
```
The second form requires that either:
1. you remember the electron's mass of the top of your head so you can directly enter it 
```python
thermal_speed(T, 9.1094e-28 * u.g)
```
or

2. remember that the thermal speed depends only on temperature and mass, and write
```python
thermal_speed(T, Particle("e-").mass)
```

Not that option (2) is hard to remember, but the point of the package is so that you don't even *need* to remember.

And for those few times where you do want to use a specific mass, it's pretty easy to write 
```python
thermal_speed(T, CustomParticle(mass=5. * u.g))
```

Reply:
With respect to the decorator...one possibility would be to have `@particle_input` pass through a `CustomParticle` when supplied with a mass or charge.

Reply:
Ideally, I'd like us to be able to do both

```python
thermal_speed(5*u.eV, "e")
thermal_speed(5 * u.eV, 5. * u.g)
```

With that said, I'll back that

```python
thermal_speed(5*u.eV, "e")
thermal_speed(5 * u.eV, CustomParticle(mass=5. * u.g))
```

is the best option for now, where we drop the `mass` kwarg altogether and just use `particle`.  I still think we should try to achieve the ideal case in the future with some kind of decorator.

Reply:
> With respect to the decorator...one possibility would be to have `@particle_input` pass through a `CustomParticle` when supplied with a mass or charge.

That could be a possibility, but the decorator might be doing too many things at that point.  Let's see when/if we get there.